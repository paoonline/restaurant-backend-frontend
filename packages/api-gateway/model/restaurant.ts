interface IRestaurant {
  name: string;
  id: number;
  coverImage: string;
  menus: string[];
  activeTimePeriod: {
    open: string;
    close: string;
  };
}

interface IShortMenu {
  name: string;
  id: string;
  thumbnailImage?: string;
  fullPrice: number;
  discountedPercent: number;
  discountedTimePeriod?: {
    begin: string;
    end: string;
  };
  sold: number;
  totalInStock: number;
}

interface IFullMenu {
  name: string;
  id: string;
  thumbnailImage?: string;
  fullPrice: number;
  discountedPercent: number;
  discountedTimePeriod?: {
    begin: string;
    end: string;
  };
  sold: number;
  totalInStock: number;
  largeImage?: string;
  options: {
    label: string;
    choices: {
      label: string;
    }[];
  }[];
}
