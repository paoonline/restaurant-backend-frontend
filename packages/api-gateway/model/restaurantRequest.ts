interface IRestaurantRequest {
    restaurantId: string
}

interface IMenuRequest {
    restaurantId: string
    menuName: string
}