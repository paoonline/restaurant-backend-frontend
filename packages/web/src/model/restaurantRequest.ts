export interface IRestaurantRequest {
    restaurantId: string
}

export interface IMenuRequest {
    restaurantId: number
    menuName: string
}