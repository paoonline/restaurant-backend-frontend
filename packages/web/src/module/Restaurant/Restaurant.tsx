import { useEffect, useMemo, useState } from "react";
import {
  restaurantService,
  restaurantShortMenuService,
} from "../../service/restaurantService";
import { Box } from "@mui/material";
import { IRestaurant, IShortMenu } from "../../model/restaurant";

import LoadingRender from "../../base/Loading";
import PaginationModule from "./module/PaginationModule";
import Banner from "./module/Banner";
import RestaurantName from "./module/RestaurantName";
import MenuLists from "./module/MenuLists";
import { useAppContext } from "../../App";
import MenuDetail from "./module/MenuDetail";
import { translations } from "../../base/lang";

const Restaurant = () => {
  const [data, setData] = useState<IRestaurant>();
  const [dataSearch, setDataSearch] = useState<string[]>([]);
  const [menuDetailLists, setMenuDetailLists] = useState<IShortMenu[]>([]);
  const [page, setPage] = useState<number>(0);
  const [openModal, setOpenModal] = useState<boolean>(false);
  const [menuName, setMenuName] = useState<string>("");
  const [isError, setError] = useState<boolean>(false)

  const { restaurantId, setLoading, loading } = useAppContext();

  const fetchData = async () => {
    setLoading(true);
    try {
      const result = await restaurantService(restaurantId);
      setData(result);
      handleMenuDetail(result);
    } catch (error) {
      setLoading(false);
      setError(true)
    } finally {
      setLoading(false);
    }
  };

  const handleMenuDetail = async (data: IRestaurant) => {
    if (page > 0) {
      setLoading(true);
    }
    try {
      const results = await Promise.all(
        data?.menus
          .filter((_, i) => handlFilterList(i))
          .map(async (element) => {
            return await restaurantShortMenuService({
              restaurantId,
              menuName: element,
            });
          })
      );

      setMenuDetailLists(results);
      setLoading(false);
    } catch (error) {
      setLoading(false);
    } finally {
      setLoading(false);
    }
  };

  const handlFilterList = (index: number) => {
    if (page === 0) {
      return index < 10;
    }
    const startPage = (page - 1) * 10;
    const endPage = (page - 1) * 10 + 10;
    return index >= startPage && index < endPage;
  };

  const handlePageChange = (_: React.ChangeEvent<unknown>, value: number) => {
    setPage(value);
  };

  const handleOnSearch = (searchValue: string) => {
    const filterSearch =
      data?.menus.filter((res) =>
        res.toLowerCase().includes(searchValue.toLowerCase())
      ) || [];
    setDataSearch(filterSearch);
  };

  const handlModalCallback = (open: boolean, id: string) => {
    setOpenModal(open);
    setMenuName(id);
  };

  const onClosedModal = (open: boolean) => {
    setOpenModal(open);
    setMenuName("");
  };

  const isClosed = useMemo(() => {
    const now = new Date();
    const currentTime = now.getHours() + now.getMinutes() / 60;

    const closeTime = data?.activeTimePeriod?.close
      ? parseFloat(data.activeTimePeriod.close)
      : 0;

    const openTime = data?.activeTimePeriod?.open
      ? parseFloat(data.activeTimePeriod.open)
      : 0;

    return currentTime >= closeTime || currentTime <= openTime;
  }, [data?.activeTimePeriod.close, data?.activeTimePeriod.open]);

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    page > 0 && data && handleMenuDetail(data);
  }, [page]);

  return (
    <Box data-testid="restautant">
      <>
        {data ? (
          <>
            <Banner url={data?.coverImage} />
            <RestaurantName
              name={data?.name}
              isClosed={isClosed}
              onSearch={handleOnSearch}
              onOpenModal={handlModalCallback}
              searchLists={dataSearch}
            />
            <MenuLists
              onOpenModal={handlModalCallback}
              data={data}
              callback={handlFilterList}
              menuDetailLists={menuDetailLists}
            />
            {openModal && <MenuDetail
              open={openModal}
              menuName={menuName}
              onClosed={onClosedModal}
            />}
            <PaginationModule
              callback={handlePageChange}
              totalPage={Math.floor(data?.menus.length / 10)}
            />
          </>
        ) : isError && <p data-testid="error_not_found">{translations.seach_not_found}</p>}
        {loading && <LoadingRender />}
      </>
    </Box>
  );
};

export default Restaurant;
