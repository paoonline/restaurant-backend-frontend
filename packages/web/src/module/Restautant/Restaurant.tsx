import { useEffect, useMemo, useState } from "react";
import {
  restaurantService,
  restaurantShortMenuService,
} from "../../service/restaurantService";
import { Box } from "@mui/material";
import { IRestaurant, IShortMenu } from "../../model/restaurant";

import LoadingRender from "../../base/Loading";
import PaginationModule from "./module/PaginationModule";
import Banner from "./module/Banner";
import RestaurantName from "./module/RestaurantName";
import MenuLists from "./module/MenuLists";
import RestaurantModal from "./module/RestaurantModal";

const Restaurant = ({ shopId }: { shopId: number }) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [data, setData] = useState<IRestaurant>();
  const [dataSearch, setDataSearch] = useState<string[]>([]);
  const [menuDetailLists, setMenuDetailLists] = useState<IShortMenu[]>([]);
  const [page, setPage] = useState<number>(0);
  const [openModal, setOpenModal] = useState<boolean>(false);

  const isClosed = useMemo(() => {
    const now = new Date();
    const currentTime = now.getHours() + now.getMinutes() / 60;

    const closeTime = data?.activeTimePeriod?.close
      ? parseFloat(data.activeTimePeriod.close)
      : 0;

    const openTime = data?.activeTimePeriod?.open
      ? parseFloat(data.activeTimePeriod.open)
      : 0;

    return currentTime >= closeTime || currentTime <= openTime;
  }, [data?.activeTimePeriod.close, data?.activeTimePeriod.open]);

  const fetchData = async () => {
    setLoading(true);
    try {
      const result = await restaurantService(shopId);
      setData(result);
      handleMenuDetail(result);
    } catch (error) {
      setLoading(false);
      console.log(error);
    }
  };

  const handleMenuDetail = async (data: IRestaurant) => {
    if (page > 0) {
      setLoading(true);
    }
    try {
      const results = await Promise.all(
        data.menus
          .filter((_, i) => handlFilterList(i))
          .map(async (element) => {
            return await restaurantShortMenuService({
              restaurantId: shopId,
              menuName: element,
            });
          })
      );
      setMenuDetailLists(results);
      setLoading(false);
    } catch (error) {
      setLoading(false);
    }
  };

  const handlFilterList = (index: number) => {
    const startPage = page * 10;
    const endPage = page * 10 + 10;

    if (page === 0) {
      return index < 10;
    }
    return index >= startPage && index < endPage;
  };

  const handlePageChange = (_: React.ChangeEvent<unknown>, value: number) => {
    setPage(value);
  };

  const handleOnSearch = (searchValue: string) => {
    const filterSearch =
      data?.menus.filter((res) =>
        res.toLowerCase().includes(searchValue.toLowerCase())
      ) || [];
    setDataSearch(filterSearch);
  };

  const handlModal = (open: boolean, id?: string) => {
    setOpenModal(open);
    id && console.log(id)
  };

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    page > 0 && data && handleMenuDetail(data);
  }, [page]);

  return (
    <Box>
      <>
        {!!data && (
          <>
            <Banner url={data?.coverImage} />
            <RestaurantName
              name={data?.name}
              isClosed={isClosed}
              onSearch={handleOnSearch}
              onOpenModal={handlModal}
              searchLists={dataSearch}
            />
            <MenuLists
              onOpenModal={handlModal}
              data={data}
              callback={handlFilterList}
              menuDetailLists={menuDetailLists}
            />
            <RestaurantModal open={openModal} onOpen={handlModal}/>

            <PaginationModule
              callback={handlePageChange}
              totalPage={Math.floor(data?.menus.length / 10)}
            />
          </>
        )}

        {loading && <LoadingRender />}
      </>
    </Box>
  );
};

export default Restaurant;
