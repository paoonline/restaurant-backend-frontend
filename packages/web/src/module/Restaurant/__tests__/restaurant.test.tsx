import React from "react";
import {
  act,
  fireEvent,
  render,
  waitFor,
  screen,
} from "@testing-library/react";
import Restaurant from "../Restaurant";
import { IFullMenu, IRestaurant, IShortMenu } from "../../../model/restaurant";
import axios from "axios";
jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

const mockRestaurant: IRestaurant = {
  name: "Mock Restaurant",
  id: 123,
  coverImage: "https://example.com/image.jpg",
  menus: ["Menu 1"],
  activeTimePeriod: {
    open: "08:00",
    close: "22:00",
  },
};

const mockShortMenu: IShortMenu = {
  name: "Menu 1",
  id: "Menu 1",
  thumbnailImage: "https://example.com/spaghetti-thumbnail.jpg",
  fullPrice: 12.99,
  discountedPercent: 20,
  discountedTimePeriod: {
    begin: "2025-01-01",
    end: "2025-01-31",
  },
  sold: 100,
  totalInStock: 50,
};

const mockFullMenu: IFullMenu = {
    name: "Special Menu",
    id: "menu123",
    thumbnailImage: "https://example.com/thumbnail.jpg", // Optional field
    fullPrice: 1500,
    discountedPercent: 20,
    discountedTimePeriod: {
      begin: "2025-01-01T00:00:00Z",
      end: "2025-12-31T23:59:59Z",
    },
    sold: 500,
    totalInStock: 200,
    largeImage: "https://example.com/large-image.jpg", // Optional field
    options: [
      {
        label: "Size",
        choices: [
          { label: "Small" },
          { label: "Medium" },
          { label: "Large" },
        ],
      },
      {
        label: "Topping",
        choices: [
          { label: "Cheese" },
          { label: "Bacon" },
          { label: "Avocado" },
        ],
      },
    ],
  };

describe("Restaurant", () => {
  it("should render correctly", () => {
    act(() => {
      mockedAxios.get.mockResolvedValueOnce({ data: mockRestaurant });
    });
    const { getByTestId } = render(<Restaurant />);
    const selectElement = getByTestId("restautant");
    expect(selectElement).not.toBeNull();
  });

  it("should render error screen", async () => {
    act(() => {
      mockedAxios.get.mockRejectedValueOnce({
        response: {
          status: 500,
          data: { message: "Internal Server Error" },
        },
      });
    });
    render(<Restaurant />);
    expect(await screen.findByText("No results found")).not.toBeNull();
  });

  it("should display image correctly", async () => {
    act(() => {
      mockedAxios.get.mockResolvedValueOnce({ data: mockRestaurant });
    });

    waitFor(() => {
        mockedAxios.get.mockResolvedValueOnce({ data: mockShortMenu });
    });
    const { findByTestId } = render(<Restaurant />);
    const selectElement = await findByTestId("image0")

    expect(selectElement).not.toBeNull();
  });

  it("should search correctly", async () => {
    act(() => {
      mockedAxios.get.mockResolvedValueOnce({ data: mockRestaurant });
    });

    waitFor(() => {
        mockedAxios.get.mockResolvedValueOnce({ data: mockShortMenu });
    });
    const { findByTestId } = render(<Restaurant />);
    const selectElement = (await findByTestId("restaurant-search")).querySelector('input')

    act(() => {
        fireEvent.change(selectElement as Element, {target: {value: 'Menu1'}})
    })

    expect(selectElement?.value).toBe('Menu1');
  });

  it("should display detail modal correctly and can closed the modal", async () => {
    waitFor(() => {
      mockedAxios.get.mockResolvedValueOnce({ data: mockRestaurant });
    });

    waitFor(() => {
        mockedAxios.get.mockResolvedValueOnce({ data: mockShortMenu });
    });

    waitFor(() => {
        mockedAxios.get.mockResolvedValueOnce({ data: mockFullMenu });
    });
    const { findByTestId } = render(<Restaurant />);
    const selectElement = await findByTestId("menu-0")

    act(() => {
        fireEvent.click(selectElement)
    })

    const selectModalElement = await findByTestId("menu-detail-modal")
    expect(selectModalElement).not.toBeNull()

    const closeButton =  (await findByTestId("menu-detail-modal-closed"))

    act(() => {
        fireEvent.click(closeButton as Element)
        
    })

    waitFor(() => {
        expect(selectModalElement).toBeNull()
    })

  });
});