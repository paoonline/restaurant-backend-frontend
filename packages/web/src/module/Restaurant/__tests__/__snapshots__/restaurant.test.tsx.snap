// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaurant should display detail modal correctly 1`] = `
<div
  aria-describedby="keep-mounted-modal-description"
  aria-labelledby="keep-mounted-modal-title"
  class="MuiModal-root css-1g04pbp-MuiModal-root"
  data-testid="menu-detail-modal"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-4nmryk-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiBox-root css-fcb265"
    tabindex="-1"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
      id="keep-mounted-modal-title"
    >
      No results found
    </h2>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
